src/types.rs

```diff
@@ -204,13 +204,21 @@ impl Types {
                     Term::Bool(pol).into(),
                 );
 
-                h.insert(i.clone(), inst_var);
+                let inst_tail = RichTerm::app(
+                    RichTerm::app(
+                        RichTerm::var("forall_tail".to_string()),
```

Comment:
What's happening here? I can't quite follow.

Reply:
So, `h` was a hashtable mapping type variables `a, b, ...` in the environment to the corresponding polymorphic contracts. Said contracts are built with the appropriate information when entering a `forall`, so that when encountering a type variable occurrence later, one can get the corresponding contract from `h`. Now we need a second type of polymorphic contract, namely a contract for polymorphic tails `{ ... | a}` to be available, so it is added to this table. Rereading this it's a bit silly to build these contracts in advance just to clone them when they are needed. I should just store the polarity and the unique identifier `sy`, which are enough to generate corresponding contracts when needed. Let's keep it like that in this PR: it will be taken care of in a related coming cleaning PR (which goal is to address [your remark](https://github.com/tweag/nickel/pull/159#discussion_r502206850)).